src/PostgREST/Config.hs

```diff
@@ -496,15 +497,25 @@ readPGRSTEnvironment =
 --
 -- >>> addFallbackAppName strangeVer "postgres:///postgres?host=server&port=5432"
 -- "postgres:///postgres?host=server&port=5432&fallback_application_name=PostgREST%2011%271%260%40%23%24%25%2C.%3A%22%5B%5D%7B%7D%3F%2B%5E%28%29%3Dasdfqwer"
+--
+-- Special chars in password
+-- >>> addFallbackAppName ver "postgres:///postgres:pass=()[]&#?host=server&port=5432"
+-- "postgres:///postgres:pass=()[]&#?host=server&port=5432&fallback_application_name=PostgREST%2011.1.0%20%285a04ec7%29"
 addFallbackAppName :: ByteString -> Text -> Text
 addFallbackAppName version dbUri = dbUri <>
-  case uriQuery <$> parseURI (toS dbUri) of
-    Nothing  -> " " <> keyValFmt -- Assume key/value connection string if the uri is not valid
-    Just ""  -> "?" <> uriFmt
-    Just "?" -> uriFmt
-    _        -> "&" <> uriFmt
+  if isPgURI dbUri
+    then case snd $ T.breakOn "?" dbUri of
```

Comment:
Does it work when the password has a `?` character inside? It works OK when it's present in `v11.0.1`.

Reply:
On main I get:

```
ghci> addFallbackAppName "11.0.0" "postgres:///postgres:pass=qwer?xyz"
"postgres:///postgres:pass=qwer?xyz&fallback_application_name=PostgREST%2011.0.0"
```

Which seems to be broken? Or does the `&` work without needing the `?` first?

---

On this branch it gives the same result:

```
ghci> addFallbackAppName "11.0.0" "postgres:///postgres:pass=qwer?xyz"
"postgres:///postgres:pass=qwer?xyz&fallback_application_name=PostgREST%2011.0.0"
```

Reply:
Yeah, just confirmed that works on v11.0.1 and also works for:

```
PGRST_DB_URI='postgres://id:mysecret?password@localhost:5432/postgres?application_name=postgrest' postgrest-run
```

And the `application_name` does get applied to the connection (confirmed by querying pg_Stat_activity)

---

The libpq parsing for the password follows the condition that it is contained between `:` and `@`:

https://github.com/postgres/postgres/blob/a829b704015104f49a11ea007957ef03a03dc0d4/src/interfaces/libpq/fe-connect.c#L6522-L6549

I think we can do something similar here.

Reply:
Or we might do the simplest thing: refuse to append fallback app name if the uri has more than one `?` character.

Reply:
We can document that as: "make sure you're URI is correctly url-encoded to ensure the fallback_application_name gets added".

Reply:
This should be good enough until we have a good connstring parser: https://github.com/jfischoff/postgres-options/issues/8